src/pytorch_lightning/strategies/ipu.py

```diff
@@ -272,6 +255,15 @@ def to_tensor(x):
         args = apply_to_collection(args, dtype=(int, float), function=to_tensor)
         return args
 
+    @staticmethod
+    def batch_to(data: Tensor) -> Tensor:
+        return data.half()
+
+    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
+        if self.precision_plugin.precision in (PrecisionType.MIXED, PrecisionType.HALF):
+            batch = apply_to_collection(batch, (FloatTensor, torch.cuda.FloatTensor), function=self.batch_to)
+        return batch
```

Comment:
since this wasn't overridden before, was it sending it to the IPU device before using the super class method?

Reply:
also, now at what point does it send the batch to the IPU device?

Reply:
> since this wasn't overridden before, was it sending it to the IPU device before using the super class method?

This used to happen in the forward pass in the `LightningIPUModule`, which was too late. Because by this time the data was already on the IPU.

> also, now at what point does it send the batch to the IPU device?

This is handled by PopTorch via [StreamCopies](https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/user-guide.html#streamcopy) and shouldn't be done explicitly using `data.to(device)`

Reply:
so I think we can keep the `LightningIPUModule` and override `batch_to_device` to just return the batch and not do anything. This way it will be cast to fp16 first and then moved to the device inside the `forward` method.

also can you add a test?

Reply:
It can't be done in `forward` because `forward` is traced and therefore all operations in forward are compiled and uploaded to the IPU.

Would a test that checks the type of the input to `LightningIPUModule.forward` be good?

Reply:
I think yet, or if there's any existing test, you can add some assertions there.

Reply:
> The data.to(device) that is currently called doesn't actually do anything

that means `self.device`doesn't point to the actual IPU device?

Reply:
> that means `self.device` doesn't point to the actual IPU device?

I'm not sure what is supposed to have `device` as an attribute (it's not an attribute of the `IPUStrategy`), but the `device` argument of `batch_to_device` is `None` when running my reproducer.

Reply:
`self.device` is a general attribute that has the value of current device. Gives the ability to users to create tensors anytime and move them to device even if they are not coming from the dataloaders. That means this is not possible when using IPUs, right?

one can't do
```py
def training_step(...):
    some_tensor = torch.randn(...).to(self.device)  # this won't move the newly created tensor to IPU?
```

Reply:
> That means this is not possible when using IPUs, right?

That is correct